extensions/ql-vscode/src/view/model-editor/ModelTypeTextbox.tsx

```diff
@@ -0,0 +1,71 @@
+import type { ChangeEvent } from "react";
+import { useCallback, useEffect, useMemo, useState } from "react";
+import type { ModeledMethod } from "../../model-editor/modeled-method";
+import { VSCodeTextField } from "@vscode/webview-ui-toolkit/react";
+import { useDebounceCallback } from "../common/useDebounceCallback";
+
+type Props = {
+  modeledMethod: ModeledMethod | undefined;
+  typeInfo: "path" | "relatedTypeName";
+  onChange: (modeledMethod: ModeledMethod) => void;
+};
+
+export const ModelTypeTextbox = ({
+  modeledMethod,
+  typeInfo,
+  onChange,
+}: Props): JSX.Element => {
+  const enabled = useMemo(
+    () => modeledMethod && modeledMethod.type === "type",
+    [modeledMethod],
+  );
+  const [value, setValue] = useState<string | undefined>(
+    modeledMethod && modeledMethod.type === "type"
+      ? typeInfo === "path"
+        ? modeledMethod.path
+        : modeledMethod.relatedTypeName
+      : undefined,
```

Comment:
I think this can be simplified since the `typeInfo` is a union:

```suggestion
    modeledMethod[typeInfo],
```